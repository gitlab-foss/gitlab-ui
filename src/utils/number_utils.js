/**
 * Adds two numbers together
 * @param {Number} a
 * @param {Number} b
 */
export const addition = (a, b) => a + b;

/**
 * Returns the sum of all arguments
 * @param  {...Number} numbers
 */
export const sum = (...numbers) => numbers.reduce(addition);

/**
 * Returns the average of all arguments
 * @param  {...Number} numbers
 */
export const average = (...numbers) => sum(...numbers) / numbers.length;

/**
 * Convert number to engineering format, using SI suffix
 * @param {Number|String} value - Number or Number-convertible String
 * @param {Number} precision
 * @return {String} number, possibly with a suffix
 */
export const engineeringNotation = (value, precision = 2) => {
  const invalidValues = [NaN, Infinity, -Infinity];
  const num = Number(value);

  if (invalidValues.includes(num) || invalidValues.includes(Number(precision))) {
    return num.toString();
  }

  const allYourBase = {
    '-24': 'y',
    '-21': 'z',
    '-18': 'a',
    '-15': 'f',
    '-12': 'p',
    '-9': 'n',
    '-6': 'Î¼',
    '-3': 'm',
    0: '',
    3: 'k',
    6: 'M',
    9: 'G',
    12: 'T',
    15: 'P',
    18: 'E',
    21: 'Z',
    24: 'Y',
  };

  const exponentialNotation = num.toExponential(precision);

  const power = exponentialNotation.split('e').map(Number)[1] || 0;

  if (power < -24 || power > 24) {
    return exponentialNotation;
  }

  const scaledPower = 3 * Math.floor(power / 3);
  const scaledMantissa = (exponentialNotation / 10 ** scaledPower)
    // strip trailing decimals from floating point rounding errors
    .toFixed(precision)
    // strip trailing 0s after a decimal point
    .replace(/\.([1-9]*)0+$/, (_, fractionalDigits) => {
      if (fractionalDigits) {
        return `.${fractionalDigits}`;
      }
      return '';
    });

  return `${scaledMantissa}${allYourBase[scaledPower]}`;
};
