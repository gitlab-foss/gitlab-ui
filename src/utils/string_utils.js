/**
 * Split the given string after each occurrence of each of the given symbols.
 *
 * Symbols are strings, and can be of length one or more. Zero-length symbols
 * are ignored.
 *
 * Unlike with `String::split`, the symbol is left in results, with
 * the split occurring _after_ the symbol.
 *
 * For example:
 *
 *     splitAfterSymbols(['/'], 'a/b/c')    // ['a/', 'b/', 'c']
 *     splitAfterSymbols(['foo'], 'foobar') // ['foo', 'bar']
 *
 * @param {string[]} symbols The symbols to split the string by.
 * @param {string} string The string to split.
 * @returns {string[]} The resulting strings.
 */
export const splitAfterSymbols = (symbols, string) => {
  const textParts = [];
  let textPartStartIndex = 0;

  if (string.length === 0) {
    return [string];
  }

  for (let i = 0; i < string.length; ) {
    let symbolFound = false;

    for (let j = 0; j < symbols.length; j += 1) {
      const symbol = symbols[j];

      if (!symbol) {
        // eslint-disable-next-line no-continue
        continue;
      }

      symbolFound = string.slice(i, i + symbol.length) === symbol;

      if (symbolFound) {
        const textPartEndIndex = i + symbol.length;
        const textPart = string.slice(textPartStartIndex, textPartEndIndex);
        textParts.push(textPart);
        textPartStartIndex = textPartEndIndex;
        i = textPartStartIndex;
        break;
      }
    }

    if (!symbolFound) {
      i += 1;
    }
  }

  const final = string.slice(textPartStartIndex);
  if (final) {
    textParts.push(final);
  }

  return textParts;
};

export const getAvatarChar = (name) => {
  if (name) {
    // Check if first character is an emjoi
    const match = name.match(/^\p{Emoji}/u);
    if (match) {
      // Return the first match
      return match[0];
    }
    return name.charAt(0).toUpperCase();
  }

  return '';
};
