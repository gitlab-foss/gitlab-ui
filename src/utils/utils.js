import { labelColorOptions, focusableTags } from './constants';

export function debounceByAnimationFrame(fn) {
  let requestId;

  return function debounced(...args) {
    if (requestId) {
      window.cancelAnimationFrame(requestId);
    }
    requestId = window.requestAnimationFrame(() => fn.apply(this, args));
  };
}

export function throttle(fn) {
  let frameId = null;

  return (...args) => {
    if (frameId) {
      return;
    }

    frameId = window.requestAnimationFrame(() => {
      fn(...args);
      frameId = null;
    });
  };
}

export function rgbFromHex(hex) {
  const cleanHex = hex.replace('#', '');
  const rgb =
    cleanHex.length === 3
      ? cleanHex.split('').map((val) => val + val)
      : cleanHex.match(/[\da-f]{2}/gi);
  const [r, g, b] = rgb.map((val) => parseInt(val, 16));
  return [r, g, b];
}

export function rgbFromString(color, sub) {
  const rgb = color.substring(sub, color.length - 1).split(',');
  const [r, g, b] = rgb.map((i) => parseInt(i, 10));
  return [r, g, b];
}

export function hexToRgba(hex, opacity = 1) {
  const [r, g, b] = rgbFromHex(hex);

  return `rgba(${r}, ${g}, ${b}, ${opacity})`;
}

export function colorFromBackground(backgroundColor) {
  let r;
  let g;
  let b;

  if (backgroundColor.startsWith('#')) {
    [r, g, b] = rgbFromHex(backgroundColor);
  } else if (backgroundColor.startsWith('rgba(')) {
    [r, g, b] = rgbFromString(backgroundColor, 5);
  } else if (backgroundColor.startsWith('rgb(')) {
    [r, g, b] = rgbFromString(backgroundColor, 4);
  }

  if (r + g + b <= 500) {
    return labelColorOptions.light;
  }
  return labelColorOptions.dark;
}

export function uid() {
  return Math.random().toString(36).substring(2);
}

/**
 * Receives an element and validates that it can be focused
 * @param { HTMLElement } The element we want to validate
 * @return { boolean } Is the element focusable
 */

export function isElementFocusable(elt) {
  if (!elt) return false;

  const { tagName } = elt;

  const isValidTag = focusableTags.includes(tagName);
  const hasValidType = elt.getAttribute('type') !== 'hidden';
  const isDisabled = elt.getAttribute('disabled') === '' || elt.getAttribute('disabled');
  const hasValidZIndex = elt.getAttribute('z-index') !== '-1';
  const isInvalidAnchorTag = tagName === 'A' && !elt.getAttribute('href');

  return isValidTag && hasValidType && !isDisabled && hasValidZIndex && !isInvalidAnchorTag;
}

/**
 * Receives an array of HTML elements and focus the first one possible
 * @param { Array.<HTMLElement> } An array of element to potentially focus
 * @return { undefined }
 */

export function focusFirstFocusableElement(elts) {
  const focusableElt = elts.find((el) => isElementFocusable(el));

  if (focusableElt) focusableElt.focus();
}

/**
 * Prints a warning message to the console
 * @param {string} message An array of element to potentially focus
 */
export function deprecationWarning(message = '') {
  if (message.length) {
    if (!['test', 'production'].includes(process.env.NODE_ENV)) {
      /* eslint-disable-next-line no-console */
      console.warn(message);
    }
  }
}
