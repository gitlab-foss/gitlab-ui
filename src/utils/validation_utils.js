// Creates out of validation configurations from .documentation.js files and creates a readable string out of it
export const getValidationInfoText = (validation = {}) => {
  switch (validation.type) {
    case 'range':
      return `${validation.min}-${validation.max}`;
    default:
      return '';
  }
};
