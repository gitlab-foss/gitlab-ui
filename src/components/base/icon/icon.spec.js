import { shallowMount, createLocalVue } from '@vue/test-utils';
import Icon from './icon.vue';
import { iconSizeOptions } from '~/utils/constants';

const ICONS_PATH = '/path/to/icons.svg';
const TEST_SIZE = 8;
const TEST_NAME = 'check-circle';

const localVue = createLocalVue();

jest.mock('@gitlab/svgs/dist/icons.svg', () => '/path/to/icons.svg');

describe('Icon component', () => {
  let wrapper;
  let consoleSpy;

  const createComponent = (props) => {
    wrapper = shallowMount(Icon, {
      propsData: {
        size: TEST_SIZE,
        name: TEST_NAME,
        ...props,
      },
      localVue,
    });
  };

  const validateName = (name) => Icon.props.name.validator(name);

  afterEach(() => {
    wrapper.destroy();

    if (consoleSpy) {
      consoleSpy.mockRestore();
    }
  });

  describe('when created', () => {
    beforeEach(() => {
      createComponent();
    });

    it(`shows svg class "s${TEST_SIZE}" and path "${ICONS_PATH}#${TEST_NAME}"`, () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('size validator', () => {
    beforeEach(() => {
      consoleSpy = jest.spyOn(console, 'warn').mockImplementation();
    });

    const maxSize = Math.max(...iconSizeOptions);

    it('fails with size outside options', async () => {
      createComponent({
        size: maxSize + 10,
      });

      expect(consoleSpy).toBeCalledWith(
        `[gitlab-ui] Unexpected value '${maxSize + 10}' was provided for the icon size`
      );
    });

    it('passes with use-deprecated-sizes', () => {
      createComponent({
        size: maxSize + 10,
        useDeprecatedSizes: true,
      });

      expect(consoleSpy).not.toBeCalled();
    });

    it('passes with size in options', () => {
      createComponent({ size: maxSize });

      expect(consoleSpy).not.toBeCalled();
    });
  });

  describe('name validator', () => {
    it('fails with name that does not exist', () => {
      const badName = `${TEST_NAME}-bogus-zebra`;
      consoleSpy = jest.spyOn(console, 'warn').mockImplementation();

      expect(validateName(badName)).toBe(false);

      expect(consoleSpy).toHaveBeenCalledWith(
        `Icon '${badName}' is not a known icon of @gitlab/svgs`
      );
    });

    it('passes with name that exists', () => {
      expect(validateName(TEST_NAME)).toBe(true);
    });
  });
});
