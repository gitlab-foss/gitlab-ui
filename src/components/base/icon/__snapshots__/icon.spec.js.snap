// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component when created shows svg class "s8" and path "/path/to/icons.svg#check-circle" 1`] = `
<svg data-testid="check-circle-icon" aria-hidden="true" class="gl-icon s8">
  <use href="/path/to/icons.svg#check-circle"></use>
</svg>
`;
